File: web/src/lib/scoring.js
import { routesForListing } from "./targets.js";

// Mirrors rentwatch/scoring.py so the dashboard can re-rank live as the user
// adjusts the per-target weight sliders, staying consistent with the score the
// Python exporter and Telegram alerts use.

const NEUTRAL_COMMUTE_MINUTES = 50;

export function enrichListings(listings, targets, weights = [1, 1]) {
  const percentiles = pricePercentiles(
    listings
      .filter((listing) => listing.status === "active")
      .map((listing) => listing.price_pcm),
  );

  return listings.map((listing) => {
    const routes = routesForListing(listing, targets);
    const commutes = routes.map((route) => bestCommute(route.transit_minutes, route.cycling_minutes));
    const { score, breakdown } = balancedScore({
      commutes,
      pricePercentile: percentiles.get(listing.price_pcm),
      letAgreed: Boolean(listing.let_agreed),
      sizeSqft: listing.size_sqft,
      hasGarden: Boolean(listing.has_garden),
      hasParking: Boolean(listing.has_parking),
      fresh: listing.freshness === "new" || listing.freshness === "reduced",
      weights,
    });
    return {
      ...listing,
      id: `${listing.search_name || "search"}:${listing.canonical_key || listing.url || listing.address}`,
      routes,
      score,
      score_breakdown: breakdown,
      best_commute_minutes: bestCommuteAverage(commutes),
    };
  });
}

export function bestCommute(transit, cycling) {
  const values = [transit, cycling].filter(isNumber);
  return values.length ? Math.min(...values) : null;
}

export function bestCommuteAverage(commutes) {
  const values = commutes.filter(isNumber);
  return values.length ? Math.round(values.reduce((a, b) => a + b, 0) / values.length) : null;
}

export function pricePercentiles(prices) {
  const valid = prices.filter(isNumber).sort((a, b) => a - b);
  const map = new Map();
  const n = valid.length;
  valid.forEach((price, index) => {
    if (!map.has(price)) map.set(price, n > 1 ? index / (n - 1) : 0);
  });
  return map;
}

export function balancedScore({
  commutes = [],
  pricePercentile,
  letAgreed = false,
  sizeSqft,
  hasGarden = false,
  hasParking = false,
  fresh = false,
  weights = [1, 1],
}) {
  const normWeights = normalizeWeights(weights, commutes.length);
  const known = commutes
    .map((commute, index) => [normWeights[index], commute])
    .filter(([, commute]) => isNumber(commute));
  const missing = commutes.length - known.length;

  let effective;
  let imbalance = 0;
  if (known.length) {
    const totalWeight = known.reduce((sum, [w]) => sum + w, 0) || 1;
    const weightedAvg = known.reduce((sum, [w, c]) => sum + w * c, 0) / totalWeight;
    const values = known.map(([, c]) => c);
    const worst = Math.max(...values);
    effective = 0.6 * worst + 0.4 * weightedAvg;
    imbalance = values.length > 1 ? Math.max(...values) - Math.min(...values) : 0;
  } else {
    effective = NEUTRAL_COMMUTE_MINUTES;
  }

  const breakdown = {};
  let score = 100;

  breakdown.commute = -round1(effective * 1.4);
  score += breakdown.commute;

  breakdown.imbalance = -round1(imbalance * 0.5);
  score += breakdown.imbalance;

  if (pricePercentile !== undefined && pricePercentile !== null) {
    breakdown.price = -round1(clamp(pricePercentile, 0, 1) * 18);
    score += breakdown.price;
  }
  if (missing) {
    breakdown.missing_routes = -(missing * 8);
    score += breakdown.missing_routes;
  }
  if (letAgreed) {
    breakdown.let_agreed = -25;
    score += -25;
  }
  const amenities = (hasGarden ? 4 : 0) + (hasParking ? 3 : 0);
  if (amenities) {
    breakdown.amenities = amenities;
    score += amenities;
  }
  const sizeBonus = sizeBonusFor(sizeSqft);
  if (sizeBonus) {
    breakdown.size = sizeBonus;
    score += sizeBonus;
  }
  if (fresh) {
    breakdown.freshness = 4;
    score += 4;
  }

  return { score: Math.round(clamp(score, 0, 100)), breakdown: pruneZero(breakdown) };
}

export function median(values) {
  const numbers = values.filter(isNumber).sort((a, b) => a - b);
  if (!numbers.length) return null;
  const mid = Math.floor(numbers.length / 2);
  return numbers.length % 2 ? numbers[mid] : Math.round((numbers[mid - 1] + numbers[mid]) / 2);
}

function sizeBonusFor(sizeSqft) {
  if (!isNumber(sizeSqft)) return 0;
  return round1(clamp((sizeSqft - 450) / 75, 0, 6));
}

function normalizeWeights(weights, count) {
  const base = Array.isArray(weights) ? [...weights] : [];
  while (base.length < count) base.push(1);
  return base.slice(0, count).map((w) => Math.max(0, Number(w) || 0));
}

function pruneZero(breakdown) {
  const out = {};
  for (const [key, value] of Object.entries(breakdown)) {
    if (value) out[key] = value;
  }
  return out;
}

function isNumber(value) {
  return typeof value === "number" && Number.isFinite(value);
}

function round1(value) {
  return Math.round(value * 10) / 10;
}

function clamp(value, low, high) {
  return Math.max(low, Math.min(high, value));
}

File: rentwatch/scoring.py
"""Canonical balanced-score model shared by the exporter and notifications.

The score answers one question for a two-commuter household: *how good a base
is this home, balancing both commutes against rent and quality?* It deliberately
penalises the **worse** of the two commutes (both people must get to work) rather
than the average, so a flat that is 15 minutes for one person and 70 for the other
does not look as good as a balanced 40/40.

All functions are pure and operate on plain numbers/dicts so they can be reused
from the export pipeline (where amenity flags are derived) and from anywhere a
``Listing`` is available.
"""

from __future__ import annotations

from dataclasses import dataclass, field

# Neutral commute (minutes) assumed when a listing has no route data at all, so a
# missing-route listing scores below a known-good commute but is not zeroed out.
NEUTRAL_COMMUTE_MINUTES = 50.0


@dataclass(slots=True)
class ScoreInput:
    commutes: list[float | None]  # best (transit/cycle) minutes per target
    price_pcm: int | None = None
    price_percentile: float | None = None  # 0 = cheapest in set, 1 = priciest
    let_agreed: bool = False
    size_sqft: int | None = None
    has_garden: bool = False
    has_parking: bool = False
    fresh: bool = False
    weights: tuple[float, float] = (1.0, 1.0)


@dataclass(slots=True)
class ScoreResult:
    score: int
    breakdown: dict[str, float] = field(default_factory=dict)


def best_commute_minutes(
    transit_minutes: float | None, cycling_minutes: float | None
) -> float | None:
    values = [v for v in (transit_minutes, cycling_minutes) if _is_number(v)]
    return min(values) if values else None


def price_percentiles(prices: list[int | None]) -> dict[int, float]:
    """Map each distinct price to its percentile rank within the active set."""
    valid = sorted(p for p in prices if _is_number(p))
    if not valid:
        return {}
    n = len(valid)
    ranks: dict[int, float] = {}
    for index, price in enumerate(valid):
        # Average rank for ties keeps the mapping stable and monotonic.
        ranks.setdefault(price, index / (n - 1) if n > 1 else 0.0)
    return ranks


def balanced_score(data: ScoreInput) -> ScoreResult:
    weights = _normalize_weights(data.weights, len(data.commutes))
    known = [(w, c) for w, c in zip(weights, data.commutes) if _is_number(c)]
    missing = len(data.commutes) - len(known)

    if known:
        total_weight = sum(w for w, _ in known) or 1.0
        weighted_avg = sum(w * c for w, c in known) / total_weight
        worst = max(c for _, c in known)
        # Bottleneck-biased effective commute: the slower commute dominates.
        effective = 0.6 * worst + 0.4 * weighted_avg
        values = [c for _, c in known]
        imbalance = max(values) - min(values) if len(values) > 1 else 0.0
    else:
        effective = NEUTRAL_COMMUTE_MINUTES
        imbalance = 0.0

    breakdown: dict[str, float] = {}
    score = 100.0

    breakdown["commute"] = -round(effective * 1.4, 1)
    score += breakdown["commute"]

    breakdown["imbalance"] = -round(imbalance * 0.5, 1)
    score += breakdown["imbalance"]

    if data.price_percentile is not None:
        breakdown["price"] = -round(_clamp(data.price_percentile, 0.0, 1.0) * 18.0, 1)
        score += breakdown["price"]

    if missing:
        breakdown["missing_routes"] = -float(missing * 8)
        score += breakdown["missing_routes"]

    if data.let_agreed:
        breakdown["let_agreed"] = -25.0
        score += breakdown["let_agreed"]

    amenities = (4.0 if data.has_garden else 0.0) + (3.0 if data.has_parking else 0.0)
    if amenities:
        breakdown["amenities"] = amenities
        score += amenities

    size_bonus = _size_bonus(data.size_sqft)
    if size_bonus:
        breakdown["size"] = size_bonus
        score += size_bonus

    if data.fresh:
        breakdown["freshness"] = 4.0
        score += 4.0

    return ScoreResult(score=int(round(_clamp(score, 0.0, 100.0))), breakdown=breakdown)


def _size_bonus(size_sqft: int | None) -> float:
    if not _is_number(size_sqft):
        return 0.0
    # +0 at 450 sq ft, capped at +6 around 900 sq ft.
    return round(_clamp((size_sqft - 450) / 75.0, 0.0, 6.0), 1)


def _normalize_weights(weights: tuple[float, float], count: int) -> list[float]:
    base = list(weights) if weights else []
    while len(base) < count:
        base.append(1.0)
    return [max(0.0, float(w)) for w in base[:count]] or [1.0] * count


def _clamp(value: float, low: float, high: float) -> float:
    return max(low, min(high, value))


def _is_number(value: object) -> bool:
    return isinstance(value, (int, float)) and not isinstance(value, bool)
